### PR TITLE
st1 [1691][IMP] account_invoice_validate_send_email

### PR DESCRIPTION
**Pull request by TomohiroAbe-Quartile**
[1691](https://www.quartile.co/web#id=1691&action=771&model=project.task&view_type=form&menu_id=505)

### CODE REVIEW

File: account_invoice_validate_send_email/data/email_templates.xml [hunk 1]
```diff
@@ -4,11 +4,11 @@
         <field name="name">請求書送付のご案内</field>
         <field
             name="email_from"
-        ><![CDATA["${object.company_id.name}" <${(object.company_id.email or user.email)|safe}>]]></field>
-        <field name="reply_to">${object.company_id.email}</field>
+        ><![CDATA["${object.invoice_mail_template_id.name}" <${(object.invoice_mail_template_id.email or user.email)|safe}>]]></field>
+        <field name="reply_to">${object.invoice_mail_template_id.email}</field>
```
Comment (yostashiro) [inline account_invoice_validate_send_email/data/email_templates.xml]: @tack-yabeQuartile The changes in this file look incorrect/unnecessary. Please revert these changes for now.

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 2]
```diff
@@ -45,8 +46,8 @@ def get_mail_compose_message(self):
             {
                 "default_model": "account.invoice",
                 "default_res_id": self.ids[0],
-                "default_use_template": bool(template_id),
-                "default_template_id": template_id,
+                "default_use_template": bool(template.id),
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
                "default_use_template": bool(template),
```
I suppose this should do.

File: account_invoice_validate_send_email/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright 2021 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    invoice_mail_template_id = fields.Many2one("mail.template", string="Invoice")
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/res_company.py]:
```suggestion
    invoice_mail_template_id = fields.Many2one("mail.template", string="Invoice Email Template")
```

**Review by TomohiroAbe-Quartile — COMMENTED** [review]
@tack-yabeQuartile 
(If it duplicates with Mr. Tashiro's comment, I'm sorry.)
Please check Travis Log, account_invoice.py looks like contain incorrect syntax.

**Comment by tack-yabeQuartile** [comment]
readmeファイル、翻訳ファイルは明日着手します。今日の時点ではユニットテスト通ったところまでで上げております。

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 4]
```diff
@@ -18,19 +19,19 @@ class AccountInvoice(models.Model):
...
+        else:
+            return False
 
     @api.multi
     def get_mail_compose_message(self):
         self.ensure_one()
-        template_id = self._get_mail_template_id()
+        template = self._get_mail_template()
+        if not template:
+            raise UserError(_("company_id cannot detect"))
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]: @tack-yabeQuartile Should we raise an error here given that the process (invoice validation) is expected to be triggered by the automatic workflow most of the time?  In other words, would raising error here be of any use?  I was thinking maybe just `return {}` would do instead, in which case the invoice validation should still complete without sending out an email when there is no template found. I think that is actually fine.  What do you think?

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 5]
```diff
@@ -18,19 +19,19 @@ class AccountInvoice(models.Model):
...
-            ).id
-        except ValueError:
-            res = False
-        return res
+    def _get_mail_template(self):
+        self.ensure_one()
+        if self.company_id.invoice_mail_template_id:
+            return self.company_id.invoice_mail_template_id
+        else:
+            return False
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
        return self.company_id.invoice_mail_template_id
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: account_invoice_validate_send_email/views/res_company_views.xml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_company_form" model="ir.ui.view">
+        <field name="name">res.company.form</field>
+        <field name="model">res.company</field>
+        <field name="inherit_id" ref="base.view_company_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='website']" position="after">
```
Comment (yostashiro) [inline account_invoice_validate_send_email/views/res_company_views.xml]: @tack-yabeQuartile I suppose the field presentation is a bit broken if we insert the field here.  Can you find a better place to put it in?

File: account_invoice_validate_send_email/models/account_invoice.py (hunk 4 above)
Comment (tack-yabeQuartile) [reply account_invoice_validate_send_email/models/account_invoice.py]: @yostashiro Thank you for comments. After thinking back about auto workflow, I agree that this error might not be observed by user companies. So I'll delete this if-block and related operations.
I'd like to notify on README that an email doesn't sent when there is no template found.

Comment (yostashiro) [reply account_invoice_validate_send_email/models/account_invoice.py]: @tack-yabeQuartile We should still keep the if block or there will be a server error when no template is found. So, just return an empty dictionary if template is not available.

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 7]
```diff
@@ -18,19 +18,16 @@ class AccountInvoice(models.Model):
...
-                "account_invoice_validate_send_email.email_template_customer_invoice_validated"
-            ).id
-        except ValueError:
-            res = False
-        return res
+    def _get_mail_template(self):
+        self.ensure_one()
+        if not self.company_id.invoice_mail_template_id:
+            self.company_id.invoice_mail_template_id = {}
+        return self.company_id.invoice_mail_template_id
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py]:
```suggestion
        return self.company_id.invoice_mail_template_id
```

File: account_invoice_validate_send_email/models/account_invoice.py [hunk 8]
```diff
@@ -18,19 +18,16 @@ class AccountInvoice(models.Model):
...
 
     @api.multi
     def get_mail_compose_message(self):
         self.ensure_one()
-        template_id = self._get_mail_template_id()
+        template = self._get_mail_template()
```
Comment (yostashiro) [inline account_invoice_validate_send_email/models/account_invoice.py:28]:
```suggestion
        template = self._get_mail_template()
        if not template:
            return {}
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: account_invoice_validate_send_email/readme/CONFIGURE.rst [hunk 9]
```diff
@@ -5,3 +5,7 @@ use the functions of this module.
 Go to *Accounting > Settings > Management > Payment Terms*, open a payment term
 and select 'Not Auto-send Invoice' in case the invoices with the payment term
 should be outside the scope of the email notification despite the workflow process settings.
+
+Go to *Settings > Users > companies*, open your company information page
+and select a favorable template from 'Invoice Email Template'.
+Recommended one is 'Invoice - Send by Email' which fills company name automatically.
```
Comment (yostashiro) [inline account_invoice_validate_send_email/readme/CONFIGURE.rst]:
```suggestion
and select a template for 'Invoice Email Template'.
No email will be sent out upon invoice validation if no template is set for the company.
```

@tack-yabeQuartile Please also regenerate README.rst. https://qrtl.slab.com/posts/oca-maintainer-tools-r0ohrbgn#generating-readme-rst-with-oca-gen-addon-readme

File: account_invoice_validate_send_email/README.rst [hunk 10]
```diff
@@ -13,9 +13,9 @@ Account Invoice Validate Send Email
 .. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
     :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
     :alt: License: AGPL-3
-.. |badge3| image:: https://img.shields.io/badge/github-qrtl%2Fmi7--custom-lightgray.png?logo=github
-    :target: https://github.com/qrtl/mi7-custom/tree/10.0/account_invoice_validate_send_email
-    :alt: qrtl/mi7-custom
+.. |badge3| image:: https://img.shields.io/badge/github-qtrl%2Fmi7--custom-lightgray.png?logo=github
+    :target: https://github.com/qtrl/mi7-custom/tree/1691-imp-separate-mail_servers-for-each-company/account_invoice_validate_send_email
+    :alt: qtrl/mi7-custom
```
Comment (yostashiro) [inline account_invoice_validate_send_email/README.rst]: @tack-yabeQuartile Please fix the parameters in the README generation command - `qrtl` for the organization, `10.0` for the branch.

File: account_invoice_validate_send_email/views/res_company_views.xml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_company_form" model="ir.ui.view">
+        <field name="name">res.company.form</field>
+        <field name="model">res.company</field>
+        <field name="inherit_id" ref="base.view_company_form" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='payment_approval_seal_stamp']" position="after">
```
Comment (yostashiro) [inline account_invoice_validate_send_email/views/res_company_views.xml:8]: This field is pro_mi7_account and we prefer to reduce the dependency on the modules from the pro-xxx repos if anything.  But this module depends on pro_mi7_account anyway, therefore it's OK for now.

**Review by yostashiro — APPROVED** [review]
LGTM 👍
